### ISSUE TITLE
IamRoleStateMachineExecution - Policy statement must contain resources

### ISSUE THREAD

**Issue by spongeboy** [horike37/serverless-step-functions#146]
# This is a Bug Report

## Description

Also reported in [Serverless forums](https://forum.serverless.com/t/iamrolestatemachineexecution-policy-statement-must-contain-resources/5594)

Using version 1.7.0 of plugin caused a deployment error of
`IamRoleStateMachineExecution - Policy statement must contain resources`

We resolved the issue by downgrading back to version 1.6.1.

Apologies, but at the moment I am unable to supply a sample config, as I need strip out sensitive information.

Our step function uses parallel steps, and the same function is referenced in multiple steps.

## Environment
OS:                     linux
Node Version:           8.9.4
Serverless Version:     1.25.0

**Comment by ACSchott526** [comment]
There is a bug in the regex here: https://github.com/horike37/serverless-step-functions/commit/f55a54758f7638ebd06391c8b83e9b333d8727e2#diff-60425c3006015f918176a0be211f414bR15

The regex does not allow `*` or any numbers which are probably apart of most people's function arn's (wildcard and account id/region, respectively). As a result, the resources field is set to an empty array and CloudFormation complains.

**Comment by kabo** [comment]
We're still having this issue in `1.7.2`.

Our `Resource` line in `serverless.yml`:

```yaml
Resource: "arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${self:custom.stage}-flow161cfetchFromSystemPlus"
```

**Comment by uris77** [comment]
Has anyone solved this issue. I just bumped into it with version 1.8

**Comment by bouiboui** [comment]
Many people are still affected by this:
https://forum.serverless.com/t/iamrolestatemachineexecution-policy-statement-must-contain-resources/5594/9

**Comment by jimmyn** [comment]
I'm having the same issue. Had to revert to 1.6.1. Any update on this?

**Comment by nakamine-tsc** [comment]
I have the same issue with version >= 1.7.2. But in my case, the cause was **wrongly written resource definition (serverless.yml)** . By fixing serverless.yml, the problem was solved.

## serverless.yml (having the issue)

I dropped `stateMachines` between `stepFunctions` and `hellostepfunc`, which cause the issue:

```yaml
functions:
  hellofunc:
    handler: handler.hello

stepFunctions:
  hellostepfunc:
    Comment: "A Hello World example of the Amazon States Language using an AWS Lambda Function"
    StartAt: HelloWorld
    States:
      HelloWorld:
        Type: Task
            Resource: "arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-hellofunc"
        End: true
```

standard output:

```
...
CloudFormation - CREATE_FAILED - AWS::IAM::Role - IamRoleStateMachineExecution
CloudFormation - CREATE_FAILED - AWS::IAM::Role - IamRoleLambdaExecution
...
 Serverless Error ---------------------------------------
  An error occurred: IamRoleStateMachineExecution - Policy statement must contain resources. (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: f72192b2-dc10-11e8-ae70-8d0f09bb39be).
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
```

## serverless.yml (Successfully deployed)

Fixing serverless.yml, the deploy succeeds:

```yaml
stepFunctions:
  stateMachines:
    hellostepfunc:
      definition:
        Comment: "A Hello World example of the Amazon States Language using an AWS Lambda Function"
        StartAt: HelloWorld
        States:
          HelloWorld:
            Type: Task
            Resource: "arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-hellofunc"
            End: true
```

In my opinion, although it was obvious I was just careless, the error message I got was confusing which can make developers difficult to detect the cause.
I think it will be nice if the framework detects the broken resource definition, then the notify the developer which definition is wrong.

**Comment by horike37** [comment]
This problem would be fixed on #154 PR.
Will anyone tests whether it is fixed properly? If so, I will cut a patch release.

**Comment by StevenWolfe** [comment]
I've also encountered this error with a hello-world setup that doesn't involve an ARN.

This config is intended to create a simple step function with a single Pass step, rather than executing a lambda:

**serverless.yml**
```yaml
service: StepFnHelloWorld

provider:
  name: aws
  runtime: nodejs8.10
  
stepFunctions:
  stateMachines:
    hellostepfunc1:
      definition:
        Comment: "A Hello World example of the Amazon States Language using an AWS Lambda Function"
        StartAt: HelloWorld1
        States:
          HelloWorld1:
            Type: Pass
            End: true

plugins:
  - serverless-step-functions
```
Output from `sls deploy`:

```
Serverless: Packaging service...
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
.....
Serverless: Stack create finished...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
.....
Serverless: Operation failed!

  Serverless Error ---------------------------------------

  An error occurred: IamRoleStateMachineExecution - Policy statement must contain resources. (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: [removed]).

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.10.0
     Serverless Version:     1.33.0
```

**Comment by gekitsuu** [comment]
Still having this problem in version 1.9.0

**Comment by g-pelletier** [comment]
The problem is still there in 1.10.0.

**Comment by theburningmonk** [comment]
right, so I think the original problem was that when the state machine doesn't reference any function, it generates a policy statement:
```
{
    "Effect": "Allow",
    "Action": [
        "lambda:InvokeFunction"
    ],
    "Resource": []
}
```
which is invalid.
That was fixed in 1.10.0, but instead, we now generate an empty array for policy statements instead, and that also fails CF's validation.
```
"Policies": [
  {
    "PolicyName": "dev-us-east-1-step-functions-guide-statemachine",
    "PolicyDocument": {
      "Version": "2012-10-17",
      "Statement": []
    }
  }
]
```
What the AWS console does, when it generates an IAM role for a state machine without any Tasks, is something like this:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Deny",
            "Action": "*",
            "Resource": "*"
        }
    ]
}
```
That's easy enough to mimic.

**Comment by horike37** [comment]
This has been fixed in v1.11.0 🎉
